File: epms-api/app/schemas/pa.py
"""Pydantic schemas for Payment Application (PA)."""
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field, model_validator

from app.schemas.current_step import CurrentStep

PA_WORKFLOW = [
    {"step": 0, "role": "finance_bp",      "label": "Finance BP Review"},
    {"step": 1, "role": "finance_manager", "label": "Finance Manager Approval"},
]


# ── Line items ─────────────────────────────────────────────────────────────────

class PaLineItemIn(BaseModel):
    po_line_id: uuid.UUID | None = None
    description: str = Field(min_length=1, max_length=500)
    qty: Decimal = Field(gt=0)
    unit: str = Field(min_length=1, max_length=30)
    # unit_price may be 0 or negative, mirroring PO lines: a discount / rebate /
    # credit line carries a negative price (header subtotal stays >= 0).
    unit_price: Decimal
    notes: str | None = None

    @property
    def line_total(self) -> Decimal:
        return (self.qty * self.unit_price).quantize(Decimal("0.01"))


class PaLineItemResponse(BaseModel):
    id: uuid.UUID
    po_line_id: uuid.UUID | None
    description: str
    qty: Decimal
    unit: str
    unit_price: Decimal
    line_total: Decimal
    notes: str | None
    sort_order: int

    model_config = {"from_attributes": True}


# ── PA ────────────────────────────────────────────────────────────────────────

class PaCreate(BaseModel):
    # The PRIMARY purchase order. Kept for callers (and stored history) that
    # know only one PO; `po_ids` is the full set and always wins when given.
    po_id: uuid.UUID | None = None
    # One PA may settle several POs of the same vendor in one payment — the
    # ordinary AP case of one cheque covering three orders. Order is meaningful:
    # the first entry becomes the primary PO mirrored onto the PA header.
    po_ids: list[uuid.UUID] = Field(default_factory=list)
    # Agreement-sourced PA (no PO, no GR). Exactly one of po_id / agreement_id
    # must be set — a PA with neither is OA's Direct PA, which EPMS does not own.
    agreement_id: uuid.UUID | None = None
    title: str = Field(min_length=1, max_length=255)
    pa_type: str = Field(default="regular", max_length=20)  # regular | prepayment | settlement | balance
    prepayment_pa_id: uuid.UUID | None = None   # required for settlement / balance types
    prepayment_applied: Decimal | None = Field(default=None, ge=0)  # settlement/balance 抵扣的预付额
    invoice_ids: list[uuid.UUID] = Field(default_factory=list)
    gr_ids: list[uuid.UUID] = Field(default_factory=list)
    subtotal: Decimal = Field(ge=0)
    tax_amount: Decimal = Field(default=Decimal("0"), ge=0)
    tax_code: str | None = Field(default=None, max_length=20)
    tax_rate: Decimal | None = Field(default=None, ge=0, le=1)
    shipping_amount: Decimal = Field(default=Decimal("0"), ge=0)
    other_charges: Decimal = Field(default=Decimal("0"), ge=0)
    other_charges_note: str | None = Field(default=None, max_length=255)
    currency: str = Field(default="CAD", max_length=10)
    notes: str | None = None
    prepayment_pct: Decimal | None = Field(default=None, ge=1, le=100)
    expected_settlement_date: date | None = None
    line_items: list[PaLineItemIn] = Field(default_factory=list)
    receipt_override: bool = False
    receipt_override_reason: str | None = Field(default=None, max_length=500)

    @property
    def resolved_po_ids(self) -> list[uuid.UUID]:
        """The PA's purchase orders, de-duplicated, order preserved.

        `po_ids` wins when present; a caller that only sent `po_id` (the old
        single-PO body, and everything already in the wild) resolves to a
        one-element list, so the endpoint below has exactly one shape to reason
        about. Empty on the agreement route.
        """
        raw = self.po_ids or ([self.po_id] if self.po_id is not None else [])
        seen: set[uuid.UUID] = set()
        out: list[uuid.UUID] = []
        for pid in raw:
            if pid not in seen:
                seen.add(pid)
                out.append(pid)
        return out

    @model_validator(mode="after")
    def _exactly_one_source(self):
        # Normalize before the XOR so `po_ids=[x]` alone is as valid as
        # `po_id=x` alone, and the two agreeing is not an error.
        pos = self.po_ids or ([self.po_id] if self.po_id is not None else [])
        if (not pos) == (self.agreement_id is None):
            raise ValueError("Provide exactly one of po_id/po_ids or agreement_id")
        if pos and self.po_id is not None and self.po_id not in pos:
            raise ValueError("po_id must be one of po_ids")
        return self


class PaUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    # Replaces the PA's whole PO set (draft / returned only, PO route only).
    # None = leave the POs alone; an empty list is rejected by the endpoint —
    # a PO-route PA with no PO would become indistinguishable from OA's Direct
    # PA and drop out of every EPMS list.
    po_ids: list[uuid.UUID] | None = None
    invoice_ids: list[uuid.UUID] | None = None
    gr_ids: list[uuid.UUID] | None = None
    subtotal: Decimal | None = Field(default=None, ge=0)
    tax_amount: Decimal | None = Field(default=None, ge=0)
    tax_code: str | None = Field(default=None, max_length=20)
    tax_rate: Decimal | None = Field(default=None, ge=0, le=1)
    shipping_amount: Decimal | None = Field(default=None, ge=0)
    other_charges: Decimal | None = Field(default=None, ge=0)
    other_charges_note: str | None = None
    notes: str | None = None
    prepayment_pct: Decimal | None = Field(default=None, ge=1, le=100)
    prepayment_applied: Decimal | None = Field(default=None, ge=0)
    expected_settlement_date: date | None = None
    line_items: list[PaLineItemIn] | None = None


class PaActionRequest(BaseModel):
    action: str = Field(min_length=1, max_length=20)
    comment: str | None = None
    bank_account_id: uuid.UUID | None = None   # funding bank/card for action='process'
    # Vendor credits to net off this payment (action='process' only).
    # THREE-VALUED — compare with `is None`, exactly as finance-api's
    # PaymentExecuteRequest.credit_ids does:
    #   None  -> apply finance-api's automatic FIFO default
    #   []    -> apply nothing this run
    #   [ids] -> apply only these
    # The Process dialog sends this ONLY when the operator deselected a credit
    # the preview offered; an untouched dialog leaves it absent so the server
    # keeps choosing.
    credit_ids: list[uuid.UUID] | None = None


class PaSettleRequest(BaseModel):
    note: str | None = None
    variance: Decimal = Field(default=Decimal("0"))
    resolution: str = Field(default="settled", max_length=20)  # settled | disputed


class PaResponse(BaseModel):
    id: uuid.UUID
    pa_number: str
    title: str
    po_id: uuid.UUID | None
    po_number: str | None
    # Every PO this PA pays, primary first. Populated from pa_po_links by
    # crud.pa (see attach_po_links); [] on the agreement route. po_id/po_number
    # above stay the primary PO so older clients keep rendering something.
    po_ids: list[uuid.UUID] = Field(default_factory=list)
    po_numbers: list[str] = Field(default_factory=list)
    vendor_id: uuid.UUID
    vendor_name: str
    invoice_ids: list
    gr_ids: list
    pa_type: str
    prepayment_pa_id: uuid.UUID | None
    prepayment_applied: Decimal | None
    subtotal: Decimal
    tax_amount: Decimal
    tax_code: str | None
    tax_rate: Decimal | None
    shipping_amount: Decimal
    other_charges: Decimal
    other_charges_note: str | None
    payment_amount: Decimal
    currency: str
    status: str
    notes: str | None
    submitted_at: datetime | None
    prepayment_pct: Decimal | None
    expected_settlement_date: date | None
    settlement_status: str | None
    settled_at: datetime | None
    settled_by: uuid.UUID | None
    settled_by_name: str | None
    settlement_note: str | None
    settlement_variance: Decimal | None
    approval_step_idx: int
    created_by: uuid.UUID
    created_by_name: str | None = None
    created_at: datetime
    updated_at: datetime
    line_items: list[PaLineItemResponse]
    receipt_override: bool = False
    receipt_override_reason: str | None = None
    receipt_override_by: uuid.UUID | None = None
    current_step: CurrentStep | None = None
    agreement_id: uuid.UUID | None = None
    agreement_number: str | None = None

    model_config = {"from_attributes": True}


class PaListResponse(BaseModel):
    items: list[PaResponse]
    total: int

File: epms/src/services/pa.ts
import { api, fetchAllPages } from '@/lib/api'
import type { ApiEvent } from './pr'
import type { WorkflowNodeDef, CurrentStep } from '@/types'

export type PaStatus =
  | 'draft'
  | 'submitted'
  | 'in_review'
  | 'approved'
  | 'processed'
  | 'returned'
  | 'rejected'
  | 'cancelled'

export type PaType = 'regular' | 'prepayment' | 'settlement' | 'balance'

// Display labels for every PA type. Single source so detail/list views never
// fall back to a binary "prepayment vs Regular" check (which mislabels
// settlement/balance as "Regular Payment").
export const PA_TYPE_LABEL: Record<PaType, string> = {
  regular: 'Regular Payment',
  prepayment: 'Prepayment',
  settlement: 'Settlement',
  balance: 'Balance',
}

export type PaAction = 'submit' | 'approve' | 'return' | 'reject' | 'process' | 'cancel'

export type SettlementStatus = 'pending' | 'settled' | 'disputed'

export interface ApiPaLineItem {
  id: string
  po_line_id: string
  description: string
  qty: number
  unit: string
  unit_price: number
  line_total: number
  notes?: string
}

export interface ApiPa {
  id: string
  pa_number: string
  title: string
  pa_type: PaType
  prepayment_pa_id?: string | null
  prepayment_applied?: number | null
  status: PaStatus
  currency: string
  subtotal: number
  tax_amount: number
  tax_code?: string | null
  tax_rate?: number | null
  // Charge breakdown extras (Decimals arrive as strings — coerce with Number() at use sites).
  shipping_amount?: number | null
  other_charges?: number | null
  other_charges_note?: string | null
  payment_amount: number
  vendor_id: string
  vendor_name: string
  // Agreement-sourced PAs (Phase 1A) have no PO — the backend returns null for
  // both. Previously typed as non-nullable, which was a lie the moment
  // agreement-sourced PAs shipped (PaListPage.tsx renders po_number directly;
  // a null renders as blank, which is fine, but callers doing po_id-truthy
  // checks need the real type).
  po_id: string | null
  po_number: string | null
  // Every PO this payment settles, primary first — one PA may cover several
  // POs of the same vendor. po_id/po_number above are just the primary one, so
  // anything rendering "which PO(s)" must use these. Empty on the agreement
  // route. Optional in the type because a cached response from before the
  // multi-PO release has neither field.
  po_ids?: string[]
  po_numbers?: string[]
  // Agreement-sourced PAs (Phase 1A) carry these instead — both null for a
  // PO-sourced PA. Mirrors PaResponse.agreement_id/agreement_number
  // (epms-api/app/schemas/pa.py) — added when the agreement create route
  // (Task 10) needed to distinguish "this PA belongs to my agreement" from
  // the po_id/po_number pair, which the backend leaves null on this route.
  agreement_id?: string | null
  agreement_number?: string | null
  invoice_ids: string[]
  prepayment_pct?: number
  expected_settlement_date?: string
  settlement_status?: SettlementStatus
  settled_at?: string
  settled_by?: string
  settled_by_name?: string
  settlement_note?: string
  settlement_variance?: number
  line_items: ApiPaLineItem[]
  notes?: string
  approval_step_idx: number
  created_by_name?: string | null
  created_at: string
  updated_at: string
  current_step?: CurrentStep | null
}

export interface CreatePaBody {
  title: string
  pa_type: PaType
  currency: string
  subtotal: number
  tax_amount: number
  tax_code?: string | null
  tax_rate?: number | null
  shipping_amount?: number
  other_charges?: number
  other_charges_note?: string
  vendor_id: string
  vendor_name: string
  // Exactly one of po_id / agreement_id must be set — the backend 422s on a
  // body with neither or both (PaCreate._exactly_one_source in
  // epms-api/app/schemas/pa.py). The agreement route (Task 10) has no PO,
  // no GR, and no receipt-override fields — none of those apply there.
  po_id?: string
  // Multi-PO payment: every PO this PA settles, primary first. Sent instead of
  // po_id; the backend derives the primary from the first entry. All POs must
  // share a vendor and a currency, and the type must be 'regular'.
  po_ids?: string[]
  agreement_id?: string
  invoice_ids?: string[]
  gr_ids?: string[]
  prepayment_pa_id?: string
  prepayment_applied?: number
  prepayment_pct?: number
  expected_settlement_date?: string
  // Posted lines omit line_total — the backend recomputes it as qty × unit_price.
  line_items?: Omit<ApiPaLineItem, 'id' | 'line_total'>[]
  notes?: string
  // Receipt gate — set when a non-prepayment PA is submitted without a matched,
  // GR-backed invoice. Requires pa_override_receipt permission; backend re-checks.
  // Never sent on the agreement route (no receipt gate exists there).
  receipt_override?: boolean
  receipt_override_reason?: string | null
}

export interface UpdatePaBody {
  title?: string
  // NOTE: pa_type and currency are NOT editable — pa_type is derived from the
  // linked PO (prepaid PO → prepayment/settlement) and currency follows the PO.
  subtotal?: number
  tax_amount?: number
  tax_code?: string | null
  tax_rate?: number | null
  shipping_amount?: number
  other_charges?: number
  payment_amount?: number
  vendor_id?: string
  vendor_name?: string
  po_id?: string
  // Replaces the PA's whole PO set (draft / returned only). Omit to leave the
  // POs alone; [] is rejected — a PA must cover at least one PO.
  po_ids?: string[]
  invoice_ids?: string[]
  gr_ids?: string[]
  prepayment_applied?: number
  prepayment_pct?: number
  expected_settlement_date?: string
  // Posted lines omit line_total — the backend recomputes it as qty × unit_price.
  line_items?: Omit<ApiPaLineItem, 'id' | 'line_total'>[]
  notes?: string
}

export interface PaActionBody {
  action: PaAction
  comment?: string
  bank_account_id?: string   // funding bank/card for action='process'
  /**
   * Vendor credits to net off this payment (action='process' only).
   * THREE-VALUED — the wire contract mirrors finance-api's
   * PaymentExecuteRequest.credit_ids:
   *   omitted / undefined -> finance-api applies its automatic FIFO default
   *   []                  -> apply no credit at all this run
   *   [ids]               -> apply only these
   * Send it ONLY when the operator actually deselected a credit the preview
   * offered. Sending a full id list for an untouched dialog would work today
   * but would freeze a stale plan if a credit landed between preview and
   * confirm.
   */
  credit_ids?: string[]
}

export interface PaFilters {
  status?: PaStatus
  pa_type?: PaType
  department_id?: string
  vendor_id?: string
  po_id?: string
  search?: string
  page?: number
  page_size?: number
}

export interface PaListResponse {
  items: ApiPa[]
  total: number
}

export const paService = {
  list: async (filters?: PaFilters): Promise<PaListResponse> => {
    const raw = await api.get<PaListResponse | ApiPa[]>('/pa', filters)
    if (Array.isArray(raw)) return { items: raw, total: raw.length }
    return raw
  },

  listAll: (filters?: Omit<PaFilters, 'page' | 'page_size'>): Promise<PaListResponse> =>
    fetchAllPages((page, page_size) => paService.list({ ...filters, page, page_size })),

  get: (id: string) =>
    api.get<ApiPa>(`/pa/${id}`),

  create: (body: CreatePaBody) =>
    api.post<ApiPa>('/pa', body),

  update: (id: string, body: UpdatePaBody) =>
    api.patch<ApiPa>(`/pa/${id}`, body),

  action: (id: string, body: PaActionBody) =>
    api.post<ApiPa>(`/pa/${id}/action`, body),

  // Finance sign-off for a zero-cash (net-0) settlement with a variance (overpaid):
  // reconciles the prepayment + closes the invoice, no payment made.
  confirmSettlement: (id: string) =>
    api.post<ApiPa>(`/pa/${id}/confirm-settlement`, {}),

  events: (id: string) =>
    api.get<ApiEvent[]>(`/pa/${id}/events`),

  workflowSteps: (id: string) =>
    api.get<WorkflowNodeDef[]>(`/pa/${id}/workflow-steps`),
}
